### ISSUE TITLE
When backup job reaches max quota in B2 it never finishes, and recovery options are limited to adding quota, or deleting B2 bucket

### ISSUE THREAD

**Issue by etheling** [restic/restic#3582]
I am using Backblaze B2 as backend and a backup job that exceeds my quota in B2 never finishes ([similar to issue here](https://github.com/restic/restic/issues/3265)). When backup reaches max quota, snapshots, forget and purge commands no longer work and there appears to be no other option to recover the situation but to either buy more quota, or delete the B2 bucket entirely.

What lead to this situation was that data on backed up directory exceeded available quota on B2. After receiving alert from B2 that I've hit my quota limits I killed ongoing backup job. To recover, I planned to relocate data from backed up directories to bring backup size below B2 quota, then list snapshots,  delete/forget the latest snapshot and purge the files. Alas, this would not work (see below).

Expected/desired behavior:

[ 1. When restic encounters an error during backup, forget, prune, etc., fail and exit after n number of retries  ([similar to issue here](https://github.com/restic/restic/issues/3265)) ]
2. When running to/past max quota while using B2 backend, be able to list snapshots, and then forget/prune

I am not familiar with B2 API, so maybe part of problem lies with it - e.g. in order to list snapshots or remove files, maybe it is required to issue API calls that are blocked because of exceeded quota. 

Reproduction summary:

1. Get a free B2 account with 10GB quota (create bucket and do restic init)
2. restic -r b2:.... backup /some/big/files/totaling/more/than/10GB
3. restic -r b2:... snapshots

Should result in behavior described here.

Output of `restic version`

```
./restic version
restic 0.12.1 compiled with go1.16.6 on linux/arm

```

How did you run restic exactly?

[for reproduction, created free 10GB B2 account, where I created new private bucket xxxxxxxxxx and created application key to read/write it which values are set to B2_ACCOUNT... environment]

```
$ env | egrep "(B2|RESTIC)"
B2_ACCOUNT_ID=1............................3
B2_ACCOUNT_KEY=K..........................d
RESTIC_PASSWORD=passphrase
RESTIC_HOST=nas
RESTIC_REPOSITORY=b2:xxxxxxxxxx:restic

$ ./restic -r ${RESTIC_REPOSITORY} init
$ ./restic -r ${RESTIC_REPOSITORY} backup /directory/with/more/than/10GB/ofdata
...
...
Save(<data/011a19a1fa>) returned error, retrying after 552.330144ms: Close: b2_upload_file: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
[stuck here]
<CTRL+C>
$ ./restic snapshots
repository 9d4588b4 opened successfully, password is correct
Save(<lock/893617a1d0>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
[stuck here]
<CTRL+C>
$ ./restic forget --keep-within-daily 7d --keep-within-weekly 1m --keep-within-monthly 1y --keep-within-yearly 1y
repository 9d4588b4 opened successfully, password is correct
Save(<lock/f0aab7aa8e>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
[stuck here]
<CTRL+C>

$ ./restic forget a5596887
repository 9d4588b4 opened successfully, password is correct
Save(<lock/86b4a9932d>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
[stuck here]
<CTRL+C>
```

What backend/server/service did you use to store the repository?

Backblaze B2

Do you have any idea what may have caused this?

I assume that once you reach the max quota at B2 'b2_get_upload_url' API call gets blocked. And it looks like snapshots,forget,prune,... issue this API call and subsequently fail.

Do you have an idea how to solve the issue?

Maybe explore if there is a way to run restic snapshots|forget|prune using B2 API calls that can be executed after being above/at max quota?

**Comment by rawtaz** [comment]
Have you tried using the `--no-lock` global option?

**Comment by etheling** [comment]
`--no-lock` appears to help with snapshots,check. But _not_ with forget and prune. E.g.: 'restic --no-lock snapsnots' will report earlier successful snapshots , and 'restic --no-lock check' will report unreferenced packs from aborted backup. Which should be removable with prune, except that 'restic --no-lock prune' fails as per above/below:

```
$ ./restic --no-lock snapshots
repository faee1e1b opened successfully, password is correct
ID        Time                 Host        Tags        Paths
----------------------------------------------------------------------
....
7a30bb3b  2021-11-21 15:23:19  nas                     /Public/exports
----------------------------------------------------------------------
6 snapshots

$ ./restic --no-lock check
using temporary cache in /tmp/restic-check-cache-171740234
repository fbee1e9b opened successfully, password is correct
created new cache in /tmp/restic-check-cache-171740234
load indexes
check all packs
pack 80ce6b1c: not referenced in any index
......
pack 2f85490b: not referenced in any index
307 additional files were found in the repo, which likely contain duplicate data.
You can run `restic prune` to correct this.
check snapshots, trees and blobs
no errors were found snapshots
[0:01] 100.00%  6 / 6 snapshots
$ # same behavior with and without --max-unused 0
$ ./restic --no-lock prune --max-unused 0
repository fbee1e9b opened successfully, password is correct
Save(<lock/d6077bbbb2>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
....
<CTRL+C>
```

Also, trying to 'forget' earlier snapshots fails:

```
$ ./restic --no-lock forget 7a30bb3b
repository fbee1e9b opened successfully, password is correct
Save(<lock/2096c4f4cc>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
...
```

**Comment by rawtaz** [comment]
Currently `forget` requires an exclusive lock. I guess you could simply modify the source code (`cmd_forget.go`) to remove that temporarily and let it remove the snapshot files, if it's possible for it to do that (depends on B2).

In general, the thing is that prune *has* to be able to write files (hence upload files) as part of its work, and that's not something your B2 allows you to do currently. Prune takes a safe approach to how it prunes your repository, which means it only deletes data after it created the corresponding data in the repository.

There is [this PR](https://github.com/restic/restic/pull/3481) that you can try (allows prune to work without needing any relevant amount of free space - AT THE COST OF THE POSSIBLY CORRUPTING THE REPO IF THE PROCESS IS CANCELLED OR UNSUCCESSFUL IN ANY WAY), but you still need to first remove the snapshots so there's something to prune.

**Comment by etheling** [comment]
I tried 'MichaelEischer:recover-enospace' branch, but unfortunately it doesn't appear to help here (assuming that I used it correctly):

```
$ ./restic.rec-nospace prune --unsafe-recover-no-free-space someidasinstructed           
repository fbee1e9b opened successfully, password is correct
Fatal: must pass id 'faee1e1b2db5174bb487bd7c32cea470d20addcc973e25c4311640a847b05af4' to --unsafe-recover-no-free-space
$ # tried with and /wo --no-lock
$ ./restic.rec-nospace --no-lock prune --unsafe-recover-no-free-space faee1e1b2db5174bb487bd7c32cea470d20addcc973e25c4311640a847b05af4
repository fbee1e9b opened successfully, password is correct
Save(<lock/a8bacdb80e>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
...
```

So I think the current workaround / recovery option is to increase the B2 data cap / quota, and prune. And then depending on B2 buckets lifecycle settings, lower the quota again.

**Comment by rawtaz** [comment]
As you can see the prune command requires an (exclusive) lock and it's when it tries to save that lock file that B2 says "nope!". To work around that you could comment out this part of `cmd_prune.go`:

~~~
	lock, err := lockRepoExclusive(gopts.ctx, repo)
	defer unlockRepo(lock)
	if err != nil {
		return err
~~~

But please note that you have to make sure that nothing else is using the repository meanwhile. This is a hacky workaround.

If that still doesn't make things work, another thing you could try is to use `forget` to forget any snapshots but the last one (assuming this is fine with you), but this will only save as much space as those file takes up, which is usually just a few KB each. So it's not like it might make much of a difference. But if you're lucky it might be enough to "untrigger" B2's blocking (assuming they see that there's now a few KBs or even MBs left in your bucket and thereby lets you use the API again).

Another thing you could perhaps try is to delete the index files from the repository, but I don't know for sure if that is 100% safe to do, someone like @MichaelEischer should probably tell you if that is a proper course of action or not before you try it.

**Comment by etheling** [comment]
That did it. Followed by 'rebuild-index'. See below:

With the code snipped commented out, prune indeed clears the space in B2. But it looks that something gets broken in the process. 

(NOTE: I am testing this on a separate bucket that I created and filled for repro/debug so no harm has/would have been done to my data). 

```
$ ./restic.rec-nospace.patch --no-lock prune --unsafe-recover-no-free-space someidasinstructed  
repository 67d4e0d0 opened successfully, password is correct
Fatal: must pass id '67d4e0d047c4bf4e3dc9cb9b662bf78f7b11f6397dded2bd5cc92342c6231c82' to --unsafe-recover-no-free-space
$ ./restic.rec-nospace.patch --no-lock prune --unsafe-recover-no-free-space 67d4e0d047c4bf4e3dc9cb9b662bf78f7b11f6397dded2bd5cc92342c6231c82
repository 67d4e0d0 opened successfully, password is correct
loading indexes...
loading all snapshots...
finding data that is still in use for 4 snapshots
[0:00] 100.00%  4 / 4 snapshots
searching used packs...
collecting packs for deletion and repacking
[0:01] 100.00%  1338 / 1338 packs processed

to repack:            0 blobs / 0 B
this removes          0 blobs / 0 B
to delete:         5092 blobs / 4.873 GiB
total prune:       5092 blobs / 4.873 GiB
remaining:         2443 blobs / 2.967 GiB
unused size after prune: 0 B (0.00% of remaining size)

deleting unreferenced packs
[1:01] 100.00%  423 / 423 files deleted
deleting index files
[0:00] 100.00%  2 / 2 files deleted
removing 723 old packs
[2:12] 100.00%  723 / 723 files deleted
rebuilding index
Save(<index/13709a746c>) returned error, retrying after 552.330144ms: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
<.... 8 repeating lines removed ...>
Save(<index/13709a746c>) returned error, retrying after 15.394871241s: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
[0:50] 100.00%  615 / 615 packs processed
Fatal: Close: b2_get_upload_url: 403: Cannot upload files, storage cap exceeded. See the Caps & Alerts page to increase your cap.
$
```
At this point space has been freed at B2 ('--no-lock' no longer needed for check,snapshots, forget), but not everything is right either as 'check' still reports same missing packs as before and e.g. 'prune' errors out as per below (I didn't dare to try 'backup' to see what would happen):

```
$ ./restic snapshots
repository 67d4e0d0 opened successfully, password is correct
ID        Time                 Host        Tags        Paths
----------------------------------------------------------------------
....
564a2609  2021-11-22 20:09:39  nas                     /Public/exports
----------------------------------------------------------------------
8 snapshots
$ ./restic forget 564a2609
repository 67d4e0d0 opened successfully, password is correct
[0:00] 100.00%  1 / 1 files deleted
$ ./restic prune
repository 67d4e0d0 opened successfully, password is correct
loading indexes...
loading all snapshots...
finding data that is still in use for 3 snapshots
[0:00] 100.00%  8 / 8 snapshots
id 161a5d03d7ea54d822884627b364f69e0bae7f3a2895dc0b8d7fc06e7cd7ebf0 not found in repository
github.com/restic/restic/internal/repository.(*Repository).LoadBlob
	/restic/internal/repository/repository.go:162
github.com/restic/restic/internal/repository.(*Repository).LoadTree
	/restic/internal/repository/repository.go:728
github.com/restic/restic/internal/restic.loadTreeWorker
	/restic/internal/restic/tree_stream.go:37
github.com/restic/restic/internal/restic.StreamTrees.func1
	/restic/internal/restic/tree_stream.go:164
golang.org/x/sync/errgroup.(*Group).Go.func1
	/home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57
runtime.goexit
	/usr/local/go/src/runtime/asm_arm.s:841
$
```
It occurred to me to try 'rebuild-index' at this point and that finally brought everything back to order. As far as I can tell, everything checks out, restore works, and backup creates a new incremental snapshot. 👍 

```
$ ./restic rebuild-index
repository 67d4e0d0 opened successfully, password is correct
loading indexes...
getting pack files to read...
adding pack file to index 000fbc294f2779a296a6ebb59ffdf0ff74e6e7498da7d07aea82e70179a841d0
...
adding pack file to index fff4592621c201b3eb74815f664c5b6b74c2b8b5c8a58eb1f3bca300d5253add
reading pack files
[1:27] 100.00%  615 / 615 packs
rebuilding index
[0:02] 100.00%  615 / 615 packs processed
deleting obsolete index files
[0:00]          0 files deleted
done
$ ./restic check
using temporary cache in /tmp/restic-check-cache-338002058
repository 67d4e0d0 opened successfully, password is correct
created new cache in /tmp/restic-check-cache-338002058
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
no errors were found snapshots
[0:02] 100.00%  8 / 8 snapshots
$
```

Lastly, I was looking into deleting e.g. index or other files to free up space to be able to forget/prune. But it appears B2 is monitoring used storage somewhat lazily, and my backup job ran almost 1GB over data cap before write related B2 API calls started failing; so no dice there.

--

Thank You so much for the help here! Hopefully this will be helpful for anyone else ending up into same predicament. Thanks again, and thank You and others for all your work on restic!

**Comment by hoshinolina** [comment]
Just ran into the lock issue myself. Is there any chance we could get some sort of `--unsafe-no-lock` option?

Edit: Got out of the situation by patching out the locking code. This is with the sftp backend, writing to a Hetzner storage box.
